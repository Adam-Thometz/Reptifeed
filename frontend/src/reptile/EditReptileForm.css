.EditReptileForm {
  border: 2px solid #D5CE3B;
  width: 50vw;
  height: 60%;
  border-radius: 20px;
  margin: 0 auto;
  padding: 1%;
  background-color: #3E3E07;
  color: #D5CE3B;
  box-shadow: 10px 5px 5px rgb(4, 66, 4);
}

.EditReptileForm-form-group {
  width: 70%;
  margin: 0 auto;
  padding: 10px;
}

.EditReptileForm-label {
  width: 8%;
  margin: 0 1%;
}

.EditReptileForm-input {
  width: 70%;
  padding: 1%;
  border-radius: 5px;
  border: none;
}

.EditReptileForm-button {
  padding: 10px;
  margin: 0 10px;
  border-radius: 10px;
  border: 1px solid #D5CE3B;
  font-size: 1rem;
}

.submit {
  background-color: #AA7B08;
}

.delete {
  background-color: #d65454;
}