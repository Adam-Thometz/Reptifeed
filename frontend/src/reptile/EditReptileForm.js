import React, { useContext, useState } from "react";
import { useNavigate, useParams } from 'react-router-dom';
import ReptifeedApi from "../api";
import Alert from "../common/Alert";
import UserContext from "../utils/UserContext";

const EditReptileForm = () => {
  const { reptiles, setReptiles } = useContext(UserContext)
  const { reptileId } = useParams();
  const navigate = useNavigate();

  const reptile = reptiles.filter(r => r.id === +reptileId)[0];

  const init = {
    name: reptile.name,
    subspecies: reptile.subspecies,
    birthday: reptile.birthday,
    imgUrl: reptile.imgUrl
  };

  const [formData, setFormData] = useState(init);
  const [formMessages, setFormMessages] = useState([]);

  const handleChange = e => {
    const { name, value } = e.target;
    setFormData(data => ({ ...data, [name]: value }));
  };

  const handleSubmit = async e => {
    e.preventDefault();
    try {
      const updatedReptile = await ReptifeedApi.updateReptile(+reptileId, formData);
      setReptiles(data => [...data.filter(r => r.id !== reptile.id), updatedReptile]);
      navigate(`/users/${updatedReptile.ownerId}/reptiles`);
    } catch (errors) {
      setFormMessages(m => [...m, errors]);
    };
  };

  return (
    <div className="EditReptileForm">
      <h1>Edit your reptile</h1>
      <form onSubmit={handleSubmit}>
        <div className="EditReptileForm-form-group">
          <label className="EditReptileForm-label" htmlFor="name">Name:</label>
          <input
            className="EditReptileForm-input"
            type='text'
            id='name'
            name='name'
            value={formData.name}
            onChange={handleChange}
          />
        </div>
        <div className="EditReptileForm-form-group">
          <label className="EditReptileForm-label" htmlFor="subspecies">Subspecies:</label>
          <input
            className="EditReptileForm-input"
            type='text'
            id='subspecies'
            name='subspecies'
            value={formData.subspecies}
            onChange={handleChange}
          />
        </div>
        <div className="EditReptileForm-form-group">
          <label className="EditReptileForm-label" htmlFor="birthday">Birthday:</label>
          <input
            className="EditReptileForm-input"
            type='date'
            id='birthday'
            name='birthday'
            value={formData.birthday}
            onChange={handleChange}
          />
        </div>
        <div className="EditReptileForm-form-group">
          <label className="EditReptileForm-label" htmlFor="imgUrl">Picture:</label>
          <input
            className="EditReptileForm-input"
            type='text'
            id='imgUrl'
            name='imgUrl'
            value={formData.imgUrl}
            onChange={handleChange}
          />
        </div>
        <button className="EditReptileForm-submit" type="submit">Edit</button>
        {formMessages.length ? <Alert type="danger" messages={formMessages} /> : null }
      </form>
    </div>
  );
};

export default EditReptileForm;