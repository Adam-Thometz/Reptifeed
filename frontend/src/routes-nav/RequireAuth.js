import React, { useContext } from "react";
import Unauthorized from "../common/Unauthorized";
import UserContext from "../utils/UserContext";
import { useParams } from 'react-router-dom'

const RequireAuth = ({ children, adminOnly = false }) => {
  const { currUser } = useContext(UserContext);
  const { id } = useParams();

  if (currUser) {
    if (adminOnly) {
      if (currUser.isAdmin) {
        return children;
      } else {
        return <Unauthorized />;
      }
    }
    else {
      if (id) {
        if (currUser.isAdmin || currUser.id === +id) {
          return children;
        } else {
          return <Unauthorized />;
        }
      } else {
        if (currUser) {
          return children;
        } else {
          return <Unauthorized />;
        }
      }
    }
  } else {
    return <Unauthorized />;
  }
}

export default RequireAuth;