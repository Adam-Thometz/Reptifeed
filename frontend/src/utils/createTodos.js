
/** createTodos(reptiles, pantry)
 * 
 * Takes a user's reptiles and pantry and creates a list of things they need to do to unlock the app's full features
 * 
 * Returns an two arrays of Link components, one of important things and one of nice-to-haves
 */

function createTodos(reptiles, pantry) {
  /** Essential checks:
   * - reptileCheck: Checks if a user has reptiles
   * - vegetableOftenCheck: Checks if a user has a vegetable staple in their pantry
   * - proteinOftenCheck: Checks if a user has a protein staple in their pantry
   * - treatModeratelyCheck: Checks if a user has a treat staple in their pantry
   * - supplementCheck: Checks if a user has at least two different supplements
   */
  const essentialTodos = [];

  const reptileCheck = !!reptiles.length;
  const vegetableOftenCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'often');
  const proteinOftenCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'often');
  const treatModeratelyCheck = pantry.some(food => food.isTreat && food.frequency === 'moderately');
  const supplementCheck = (pantry.filter(food => food.type === 'supplement')).length >= 2;

  if (!reptileCheck) essentialTodos.push('Add a reptile');
  if (!vegetableOftenCheck) essentialTodos.push('Add vegetables eaten often');
  if (!proteinOftenCheck) essentialTodos.push('Add protein eaten often');
  if (!treatModeratelyCheck) essentialTodos.push('Add a treat eaten moderately');
  if (!supplementCheck) essentialTodos.push('Add at least two supplements');

  /** Nice-to-have checks:
   * - vegetableModeratelyCheck: Checks if a user has a vegetable that's eaten moderately in their pantry
   * - vegetableOccasionallyCheck: Checks if a user has a vegetable that's eaten occasionally in their pantry
   * - proteinModeratelyCheck: Checks if a user has a protein that's eaten moderately in their pantry
   * - proteinOccasionallyCheck: Checks if a user has a protein that's eaten occasionally in their pantry
   * - treatOccasionallyCheck: Checks if a user has a treat that's eaten rarely in their pantry
   */
  const niceToHaveTodos = [];

  const vegetableModeratelyCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'moderately');
  const vegetableOccasionallyCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'occasionally' && !food.isTreat);
  const proteinModeratelyCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'moderately');
  const proteinOccasionallyCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'occasionally' && !food.isTreat);
  const treatOccasionallyCheck = pantry.some(food => food.isTreat && food.frequency === 'occasionally');

  if (!vegetableModeratelyCheck) niceToHaveTodos.push('Add a vegetable eaten moderately');
  if (!vegetableOccasionallyCheck) niceToHaveTodos.push('Add a vegetable eaten occasionally');
  if (!proteinModeratelyCheck) niceToHaveTodos.push('Add a protein eaten moderately');
  if (!proteinOccasionallyCheck) niceToHaveTodos.push('Add a protein eaten occasionally');
  if (!treatOccasionallyCheck) niceToHaveTodos.push('Add a treat eaten occasionally');

  return {essentialTodos, niceToHaveTodos};
};

export default createTodos;