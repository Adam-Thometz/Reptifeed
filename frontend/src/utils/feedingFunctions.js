/** Feeding functions
 * 
 * Used to get foods from pantry that match frequencies and food type.
*/

function getFoodFromPantry(pantry, freq, type) {
  const availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === freq && food.type === type)));
  if (!availableFood.length) return `Get some more ${type}`
  const selectedIdx = Math.floor(Math.random() * availableFood.length);

  return availableFood[selectedIdx];
};

function getTreatFromPantry(pantry, freq) {
  const availableTreats = pantry.filter(food => food.isTreat && food.frequency === freq);
  if (!availableTreats.length) return "You don't have any treats";
  const selectedIdx = Math.floor(Math.random() * availableTreats.length); 
  
  return availableTreats[selectedIdx];
};

export { getFoodFromPantry, getTreatFromPantry };