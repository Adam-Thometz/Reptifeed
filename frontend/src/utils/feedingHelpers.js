/** Feeding functions
 * 
 * Used to get foods from pantry that match frequencies and food type.
 * 
 * If no foods are available for the frequency, returns foods eaten most often
*/

function getFoodFromPantry(pantry, freq, type) {
  let availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === freq && food.type === type)));
  if (!availableFood.length) {
    availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === 'often' && food.type === type)));
  };
  const selectedIdx = Math.floor(Math.random() * availableFood.length);

  return availableFood[selectedIdx];
};

function getTreatFromPantry(pantry, freq) {
  let availableTreats = pantry.filter(food => food.isTreat && food.frequency === freq);
  if (!availableTreats.length) {
    availableTreats = pantry.filter(food => food.isTreat);
  };
  const selectedIdx = Math.floor(Math.random() * availableTreats.length); 
  
  return availableTreats[selectedIdx];
};

function getNextSupplement(pantry, lastSupplement) {
  const nextSupplement = lastSupplement === 'calcium' ?
    pantry.find(food => food.name === 'multivitamin') :
    pantry.find(food => food.name === 'calcium');
  
  return nextSupplement;
}

export { getFoodFromPantry, getTreatFromPantry, getNextSupplement };