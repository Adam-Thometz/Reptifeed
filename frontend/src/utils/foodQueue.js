// foodQueue: Feed reptile based on frequency
const foodQueue = ["often", "often", "moderately", "often", "often", "moderately", "often", "often", "moderately", "occasionally"];

// treatQueue: Give reptile treats based on frequency
const treatQueue = ["moderately", "moderately", "occasionally"];

export { foodQueue, treatQueue };