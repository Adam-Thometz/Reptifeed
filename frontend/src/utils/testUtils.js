import React from "react";
import UserContext from "./UserContext";

const demoUser = {
  id: 1,
  username: 'rickSanchez',
  password: 'ilikeMorty',
  email: 'rickman@qmail.com',
  isAdmin: false
};

const demoReptile = [
  {
    id: 1,
    name: 'two crows',
    species: 'crows',
    subspecies: 'two of them',
    birthday: '2015-10-29',
    imgUrl: '',
    ownerId: 1
  }
];

const demoPantry = [
  {
    name: 'plumbus',
    type: 'everything',
    frequency: 'always',
    image: '',
    isTreat: true,
    tips: "Everyone needs a plumbus!"
  }
];

const UserProvider = ({ children, currUser = demoUser, pantry = demoPantry, reptiles = demoReptile }) => (
  <UserContext.Provider value={{currUser, pantry, reptiles}}>
    {children}
  </UserContext.Provider>
);

export { UserProvider };