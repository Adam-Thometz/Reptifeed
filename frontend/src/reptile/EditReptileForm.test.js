import React from "react";
import EditReptileForm from "./EditReptileForm";
import { render, UserProvider } from '../utils/testUtils';
import RequireAuth from "../routes-nav/RequireAuth";
import { Routes, Route } from 'react-router-dom';

test('it renders without crashing', () => {
  render(<UserProvider>
    <Routes>
      <Route path='/users/:id/reptiles/:reptileId/edit' element={<RequireAuth><EditReptileForm /></RequireAuth>} />
    </Routes>
  </UserProvider>, { initialRoutes: ['/users/1/reptiles/1/edit'] });
});

test('it matches snapshot', () => {
  const { asFragment } = render(
    <UserProvider>
      <Routes>
        <Route path='/users/:id/reptiles/:reptileId/edit' element={<RequireAuth><EditReptileForm /></RequireAuth>} />
      </Routes>
    </UserProvider>, { initialRoutes: ['/users/1/reptiles/1/edit'] });
  expect(asFragment()).toMatchSnapshot();
});