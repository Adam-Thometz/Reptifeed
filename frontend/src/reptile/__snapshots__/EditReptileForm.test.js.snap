// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="EditReptileForm"
  >
    <h1>
      Edit your reptile
    </h1>
    <form>
      <div
        class="EditReptileForm-form-group"
      >
        <label
          class="EditReptileForm-label"
          for="name"
        >
          Name:
        </label>
        <input
          class="EditReptileForm-input"
          id="name"
          name="name"
          type="text"
          value="two crows"
        />
      </div>
      <div
        class="EditReptileForm-form-group"
      >
        <label
          class="EditReptileForm-label"
          for="subspecies"
        >
          Subspecies:
        </label>
        <input
          class="EditReptileForm-input"
          id="subspecies"
          name="subspecies"
          type="text"
          value="two of them"
        />
      </div>
      <div
        class="EditReptileForm-form-group"
      >
        <label
          class="EditReptileForm-label"
          for="birthday"
        >
          Birthday:
        </label>
        <input
          class="EditReptileForm-input"
          id="birthday"
          name="birthday"
          type="date"
          value="2015-10-29"
        />
      </div>
      <div
        class="EditReptileForm-form-group"
      >
        <label
          class="EditReptileForm-label"
          for="imgUrl"
        >
          Picture:
        </label>
        <input
          class="EditReptileForm-input"
          id="imgUrl"
          name="imgUrl"
          type="text"
          value=""
        />
      </div>
      <button
        class="EditReptileForm-button submit"
        type="submit"
      >
        Edit
      </button>
      <button
        class="EditReptileForm-button delete"
      >
        Delete reptile
      </button>
    </form>
  </div>
</DocumentFragment>
`;
